playback/src/audio_backend/pulseaudio.rs

```diff
@@ -12,39 +12,36 @@ const STREAM_NAME: &str = "Spotify endpoint";
 
 pub struct PulseAudioSink {
     s: Option<Simple>,
-    ss: pulse::sample::Spec,
     device: Option<String>,
     format: AudioFormat,
+    pulse_format: pulse::sample::Format,
 }
 
 impl Open for PulseAudioSink {
     fn open(device: Option<String>, format: AudioFormat) -> Self {
-        info!("Using PulseAudio sink with format: {:?}", format);
+        let mut actual_format = format;
+        if actual_format == AudioFormat::F64 {
+            warn!("PulseAudio currently does not support F64 output, falling back to F32");
+            actual_format = AudioFormat::F32;
+        }
 
         // PulseAudio calls S24 and S24_3 different from the rest of the world
-        let pulse_format = match format {
+        let pulse_format = match actual_format {
             AudioFormat::F32 => pulse::sample::Format::FLOAT32NE,
             AudioFormat::S32 => pulse::sample::Format::S32NE,
             AudioFormat::S24 => pulse::sample::Format::S24_32NE,
             AudioFormat::S24_3 => pulse::sample::Format::S24NE,
             AudioFormat::S16 => pulse::sample::Format::S16NE,
-            _ => {
-                unimplemented!("PulseAudio currently does not support {:?} output", format)
-            }
+            _ => unreachable!(),
         };
 
-        let ss = pulse::sample::Spec {
-            format: pulse_format,
-            channels: NUM_CHANNELS,
-            rate: SAMPLE_RATE,
-        };
-        debug_assert!(ss.is_valid());
+        info!("Using PulseAudio sink with format: {:?}", actual_format);
 
         Self {
             s: None,
-            ss,
             device,
-            format,
+            format: actual_format,
```

Comment:
OK some more comments now I have more time to look at this. It should be so that you need only one of `format` or `pulse_format. Because you are depending on `write_as_bytes!` which assumes there is a `format` field, it may be easiest to just store `format` and move the `pulse::Sample::Format` logic down to `start`.

Reply:
I can do that.

Reply:
Done.